server/topic.go

```diff
@@ -32,18 +40,43 @@ func newTopic(id string) *topic {
 }
 
 // Subscribe subscribes to this topic
-func (t *topic) Subscribe(s subscriber, userID string, cancel func()) int {
+func (t *topic) Subscribe(s subscriber, visitor *visitor, cancel func()) int {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 	subscriberID := rand.Int()
 	t.subscribers[subscriberID] = &topicSubscriber{
-		userID:     userID, // May be empty
+		visitor:    visitor, // May be empty
 		subscriber: s,
 		cancel:     cancel,
 	}
 	return subscriberID
 }
 
+func (t *topic) Stale() bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.rateVisitorExpires.Before(time.Now()) {
+		t.rateVisitor = nil
+	}
+	return len(t.subscribers) == 0 && t.rateVisitor == nil
+}
+
+func (t *topic) SetRateVisitor(v *visitor) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.rateVisitor = v
+	t.rateVisitorExpires = time.Now().Add(rateVisitorExpiryDuration)
+}
+
+func (t *topic) RateVisitor() *visitor {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.rateVisitorExpires.Before(time.Now()) {
+		t.rateVisitor = nil
+	}
+	return t.rateVisitor
+}
+
```

Comment:
What happens if someone is continuously subscribed for longer than 12 hours (or `rateVisitorExpiryDuration`)? Shouldn't  the countdown only start after rateVisitor unsubscribed?

Also, I removed the expiry check in RateVisitor (ex. Billee) because it runs minutely through the manager/Stale anyway. The comparison is probably low cost though so it doesn't matter.

Reply:
> Shouldn't the countdown only start after rateVisitor unsubscribed?

I see. What about this:

- The `topic.expires` stays 0 until somebody unsubscribes
- `topic.Stale()` returns true if `len(subs) == 0 && expires has passed`

So we don't tie the topic expiry to the rate visitor at all anymore. We just make sure the topic sticks around for a few hours, even after everybody disconnects. 

I shortened the expiry time to 6 hours, which should be enough buffer time anyway. Realistically, the topic won't get cleaned up anyway, as long as messages are in the topic (see manager).

https://github.com/binwiederhier/ntfy/pull/633/commits/8eae44ea614483c912f6484cfb2295d12096fec1